### PR TITLE
Ensure APM service name gets exposed to Serilog/NLog ECS export"

### PR DESCRIPTION
**Pull request by Mpdreamz**
- Expose service information to serilog and expose it to EcsTextJsonFormatter
- Expose service information to NLog and ensure Elastic.CommonSchema.NLog sets Service information

Fixes: #134

### CODE REVIEW

File: src/Elastic.Apm.NLog/ApmServiceNodeNameLayoutRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using System.Text;
+using NLog;
+using NLog.Config;
+using NLog.LayoutRenderers;
+
+namespace Elastic.Apm.NLog;
+
+[LayoutRenderer(Name)]
+[ThreadSafe]
```
Comment (snakefoot) [inline src/Elastic.Apm.NLog/ApmServiceNodeNameLayoutRenderer.cs:9]: Please also include `[ThreadAgnostic]`. See also https://github.com/NLog/NLog/wiki/How-to-write-a-custom-layout-renderer#how-to-optimize-performance-for-the-layout-renderer

File: src/Elastic.Apm.NLog/ApmServiceVersionLayoutRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+using System.Text;
+using NLog;
+using NLog.Config;
+using NLog.LayoutRenderers;
+
+namespace Elastic.Apm.NLog;
+
+[LayoutRenderer(Name)]
+[ThreadSafe]
```
Comment (snakefoot) [inline src/Elastic.Apm.NLog/ApmServiceVersionLayoutRenderer.cs:9]: Please also include `[ThreadAgnostic]`. See also https://github.com/NLog/NLog/wiki/How-to-write-a-custom-layout-renderer#how-to-optimize-performance-for-the-layout-renderer

File: src/Elastic.Apm.NLog/ApmServiceNameLayoutRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+using System.Text;
+using NLog;
+using NLog.Config;
+using NLog.LayoutRenderers;
+
+namespace Elastic.Apm.NLog
+{
+	[LayoutRenderer(Name)]
+	[ThreadSafe]
```
Comment (snakefoot) [inline src/Elastic.Apm.NLog/ApmServiceNameLayoutRenderer.cs:13]: Please also include `[ThreadAgnostic]`. See also https://github.com/NLog/NLog/wiki/How-to-write-a-custom-layout-renderer#how-to-optimize-performance-for-the-layout-renderer

**Comment by Mpdreamz** [comment]
`Elasticsearch.Extensions.Logging` kinda does its own thing. It does not have the equivalent of Serilog/NLog enrichers that take precedence providing tracing/service information directly from the agent.

Need to look into ExternalScopeProviders in order to expose it without `Elasticsearch.Extensions.Logging` taking a dependency on the agent: 

See: https://github.com/dotnet/runtime/issues/34305

File: src/Elastic.Apm.NLog/ApmServiceNodeNameLayoutRenderer.cs (hunk 1 above)
Comment (Mpdreamz) [reply src/Elastic.Apm.NLog/ApmServiceNodeNameLayoutRenderer.cs:9]: Thanks @snakefoot should those also go on the other LayoutRenderers? (`Apm(Span|Transaction|Trace)Id`)

**Comment by snakefoot** [comment]
It can only be used when not capturing the state of the executing thread.

And I guess the TraceId is part thread execution context.


Sent from my Galaxy



-------- Original message --------
From: Martijn Laarman ***@***.***>
Date: 31/08/2022 23:22 (GMT+01:00)
To: elastic/ecs-dotnet ***@***.***>
Cc: Rolf Kristensen ***@***.***>, Mention ***@***.***>
Subject: Re: [elastic/ecs-dotnet] Ensure APM service name gets exposed to Serilog/NLog ECS export" (PR #209)


@Mpdreamz commented on this pull request.

________________________________

In src/Elastic.Apm.NLog/ApmServiceNodeNameLayoutRenderer.cs<https://github.com/elastic/ecs-dotnet/pull/209#discussion_r960037289>:

> @@ -0,0 +1,19 @@
+using System.Text;
+using NLog;
+using NLog.Config;
+using NLog.LayoutRenderers;
+
+namespace Elastic.Apm.NLog;
+
+[LayoutRenderer(Name)]
+[ThreadSafe]


Thanks @snakefoot<https://github.com/snakefoot> should those also go on the other LayoutRenderers? (Apm(Span|Transaction|Trace)Id)

**Pull request merged**